### PR TITLE
Added ability to control amount of suites / tasks within limited range

### PR DESCRIPTION
**Pull request by bryangordley**
## Description of the PR
This feature grants users the ability to control the maximum amount of suites and tasks per suite within a controlled range. The original limit was 6 for suites and 8 for tasks which remain the default in this feature, maximums stopping at 12 for suites and 20 for tasks.

Additionally, made some changes to the front-end testing / package.json to support testing in a Windows environment. This will need to be tested in Mac / Linux environments. As for tests, it appears that the whole configuration menu needs tests written which I am happy to add in another pull request (or this one if required).

### Issues Related
[Add New Task Missing After 8 Tasks](https://github.com/angrykoala/gaucho/issues/265)

------

Please, before doing the PR make sure that you are following the intructions described in [CONTRIBUTING.md](https://github.com/angrykoala/gaucho/blob/master/CONTRIBUTING.md) and you have completed the following checklist

**Checklist**
- [x] Your PR is done against the `dev` branch
- [x] Your origin branch is updated with the latest changes from `dev`
- [x] All the tests and linters are passing (`npm run test`)
- [ ] Your changes have tests

Thanks for contributing to Gaucho

### CODE REVIEW

File: app/common/app_config.js [hunk 1]
```diff
@@ -47,7 +49,7 @@ const defaultTasks = {
     }]
 };
 
-// Just for basic offuscation of the config file
+// Just for basic obfuscation of the config file
```
Comment (angrykoala) [inline app/common/app_config.js]: Good catch, thanks

File: app/renderer/components/range_form.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"use strict";
```
Comment (angrykoala) [inline app/renderer/components/range_form.js]: I like this component, easy to re-use

File: package.json [hunk 3]
```diff
@@ -33,7 +34,7 @@
     "icon": "resources/icon.png"
   },
   "scripts": {
-    "test": "istanbul cover _mocha -- --exit && npm run lint",
+    "test": "istanbul cover node_modules/mocha/bin/_mocha -- --exit && npm run lint",
```
Comment (angrykoala) [inline package.json]: I assume this is required for the mac testing

File: package.json [hunk 4]
```diff
@@ -20,6 +20,7 @@
     "cross-env": "^5.0.5",
     "electron": "~1.7.8",
     "electron-builder": "^19.41.1",
+    "electron-prebuilt": "^1.4.13",
```
Comment (angrykoala) [inline package.json]: It is really required `electron-prebuilt` in order to run Spectron in windows? I would like to avoid extra dependencies if possible

**Comment by angrykoala** [comment]
Hi @bryangordley 
Thanks for your PR.
I agree with the range of maximum tasks per suite you added. But I have a couple of questions:
* If a user create a suite with 10 tasks for example, and then the maximum tasks are moved to 8, now we have an inconsistent suite with more than the maximum tasks.
* What should we do when importing tasks from a configuration with a bigger maximum of tasks?

The suites present a different problem, the limit of suites was added because more than 8 suites will not be displayed correctly (as shown in the picture) and will make suite tabs extremely small as adding a scroll is not feasible. 

screenshot from 2017-11-25 16-22-39

After reviewing how it works with a configurable maximum, I feel that it doesn't make sense to make this configurable, it just over-complicates the usage.

If Gaucho works fine with 20 tasks per suite (that's 160 tasks in total!) I think it would be easier to simply make the hardcoded limit 20.

It may be useful, however, to add some "advanced options" with a numeric field to add any number of maximum tasks for people who actually need more than the maximum tasks, but that would be a separate issue. 
What do you think about all this?

I agree with the changes made for testing (check the questions in some comments). Also, most of Gaucho UI require proper e2e testing, including configuration. I'm happy to add an issue for configuration testing if you want to do it. I'll try to work on improving current tests with Spectron

File: package.json (hunk 3 above)
Comment (bryangordley) [reply package.json]: You are correct. I am not sure if this works with Mac or Linux build environments. I will give it a test run in a fresh Linux installation tonight.

File: app/renderer/components/range_form.js (hunk 2 above)
Comment (bryangordley) [reply app/renderer/components/range_form.js]: Thanks, I made my best attempt at following the structure of the existing components.

File: package.json (hunk 4 above)
Comment (bryangordley) [reply package.json]: Unfortunately I could not find another way to resolve the issue with Spectron on Windows. What I have discovered is that './node_modules/.bin/electron' should work properly but has a bug somewhere downstream.

I will do some more research and see if there is a way around this issue without adding dependencies.

**Comment by bryangordley** [comment]
Below is my stance on the points you brought to the table
- I agree that hard-coding the tasks to 20 will cover most use cases. I like the idea of an advanced section in the future that allows a configurable max in the form of a numeric input. 
- I will remove the suite range as it seems like this is not very useful and creates more issues than it resolves.
- Creating tests for the existing Configuration options would be a great learning experience. I understand how the testing environment works and am pretty comfortable with Spectron at this point.

Below are some questions that I have
- Would you like to keep the hard-coded values stored in AppStatus.Config per my change or move back into AppStatus?
- Would you like to retain the range_form.js component for future use?

**Comment by angrykoala** [comment]
Hi @bryangordley 
* If 20 tasks (or maybe less if possible) are enough for your use case of Gaucho, I think is just easier to hardcode that value for now rather than adding extra complexity to the current config, which is not designed for a lot of options. Keep in mind that the limit was added simply because I'm not sure how Gaucho will handle a huge amount of tasks.
* Perfect, I'm writing some E2E tests these days, trying to clean up a bit, I think it'll make your own tests easier, thanks, I need a bit of help writing tests with Spectron
* If you like the idea of having advanced configs, please, feel free to open an issue, once we define 2 or 3 advanced features, those could be added as a sub-menu of the current config
* All hard-coded values that aren't going to be changed in the config should go back to AppStatus (which, atm is the placeholder for a proper config file and vuex storage)
* Yes, I'm not a fan of having unused components, but I think we can use it very soon so I'm happy to keep that component in the code

Thanks!

**Comment by bryangordley** [comment]
@angrykoala Awesome, this is my first time contributing to a project on GitHub, I am still learning the ropes a bit. Is there a way to update this PR with the changes you have requested or should I close this PR and submit a new one?

**Comment by angrykoala** [comment]
Hi @bryangordley no worries, that's what these projects are for.

Usually, just modifying your own branch should automatically update the PR

**Comment by angrykoala** [comment]
Since 0.4.1, number of tasks per suites is increased, this PR will be closed as this should fix the issue discussed

**Pull request closed without merging**